ipld/src/selector/walk.rs

```diff
@@ -0,0 +1,208 @@
+
+impl Selector {
+    /// Walks all nodes visited (not just matched nodes) and executes callback with progress and
+    /// Ipld node. An optional link loader/ resolver is passed in to be able to traverse links.
+    pub async fn walk_all<L, F>(
+        self,
+        ipld: &Ipld,
+        resolver: Option<L>,
+        callback: F,
+    ) -> Result<(), Error>
+    where
+        F: Fn(&Progress<L>, &Ipld, VisitReason) -> Result<(), String> + Sync,
+        L: LinkResolver + Sync + Send + Clone,
+    {
+        Progress {
+            resolver,
+            path: Path::default(),
+        }
+        .walk_all(ipld, self, &callback)
+        .await
+    }
+
+    /// Walks a graph of Ipld nodes, executing the callback only on the nodes "matched".
+    /// If a resolver is passed in, links will be able to be traversed.
+    pub async fn walk_matching<L, F>(
+        self,
+        ipld: &Ipld,
+        resolver: Option<L>,
+        callback: F,
+    ) -> Result<(), Error>
+    where
+        F: Fn(&Progress<L>, &Ipld) -> Result<(), String> + Sync,
+        L: LinkResolver + Sync + Send + Clone,
+    {
+        self.walk_all(ipld, resolver, |prog, ipld, reason| -> Result<(), String> {
+            if let VisitReason::SelectionMatch = reason {
+                return callback(prog, ipld);
+            }
+            Ok(())
+        })
+        .await
+    }
+}
+
+/// Provides reason for callback in traversal for `walk_all`.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum VisitReason {
+    /// Ipld node visited was a specific match.
+    SelectionMatch,
+    /// Ipld node was visited while searching for matches.
+    SelectionCandidate,
+}
+
+#[async_trait]
+pub trait LinkResolver {
+    #[allow(unused_variables)]
+    /// Resolves a Cid link into it's respective Ipld node, if it exists.
+    async fn load_link(&self, link: &Cid) -> Result<Option<Ipld>, String> {
+        Err("load_link not implemented on the LinkResolver".into())
+    }
+}
+
+impl LinkResolver for () {}
+
+#[derive(Clone, Debug)]
+pub struct Progress<L = ()>
+where
+    L: Clone,
+{
+    resolver: Option<L>,
+    path: Path,
+}
+
+impl<L> Progress<L>
+where
+    L: LinkResolver + Sync + Send + Clone,
+{
+    /// Returns the path of the current progress
+    pub fn path(&self) -> &Path {
+        &self.path
+    }
+
+    #[async_recursion]
+    async fn walk_all<F>(
+        &mut self,
+        ipld: &Ipld,
+        selector: Selector,
+        callback: &F,
+    ) -> Result<(), Error>
+    where
+        F: Fn(&Progress<L>, &Ipld, VisitReason) -> Result<(), String> + Sync,
+    {
+        // Resolve any links transparently before traversing
+        if let Ipld::Link(cid) = ipld {
+            if let Some(resolver) = &self.resolver {
+                let mut node = resolver.load_link(cid).await.map_err(Error::Link)?;
+                while let Some(Ipld::Link(c)) = node {
+                    node = resolver.load_link(&c).await.map_err(Error::Link)?;
+                }
```

Comment:
I would probably just leave this out, assuming it's not a super common occurrence — if it links to another link, the recursive call will automatically do the right thing

Reply:
no, because if it tries to explore a field on it, it will stop traversing and return Ok silently, when it should have resolved the cid